File: src/main/java/io/github/smootheez/modrinth/ModrinthPublisher.java
package io.github.smootheez.modrinth;

import com.google.gson.reflect.*;
import io.github.smootheez.*;
import io.github.smootheez.exception.*;
import okhttp3.*;
import org.gradle.api.*;
import org.gradle.api.file.*;

import java.io.*;
import java.util.*;
import java.util.stream.*;

/**
 * Publishes mod versions to Modrinth using their official API.
 * <p>
 * This publisher is responsible for:
 * <ul>
 *     <li>Validating configuration such as release type, status, and game versions</li>
 *     <li>Fetching available game versions from Modrinth</li>
 *     <li>Building metadata for the version upload</li>
 *     <li>Uploading files and metadata via multipart requests</li>
 * </ul>
 */
public class ModrinthPublisher extends Publisher {

    /** Endpoint for uploading a new version to Modrinth. */
    private static final String UPLOAD_URL = "https://api.modrinth.com/v2/version";

    /** Endpoint for fetching all available game version tags from Modrinth. */
    private static final String GAME_VERSION_URL = "https://api.modrinth.com/v2/tag/game_version";

    /** Supported Modrinth version status values. */
    private static final Set<String> VALID_STATUS = Set.of("listed", "archived", "draft", "unlisted", "scheduled");

    /**
     * Constructs a new {@code ModrinthPublisher}.
     *
     * @param project    the Gradle project instance
     * @param extension  the plugin configuration extension
     * @param client     the HTTP client used for API communication
     */
    public ModrinthPublisher(Project project, McModPublisherExtension extension, OkHttpClient client) {
        super(project, extension, client);
    }

    /**
     * Executes the publishing process to Modrinth.
     * <p>
     * This includes:
     * <ul>
     *     <li>Validating configuration</li>
     *     <li>Fetching supported game versions from Modrinth</li>
     *     <li>Constructing metadata</li>
     *     <li>Sending the multipart request containing metadata and mod files</li>
     * </ul>
     * Logs and aborts on invalid configuration or missing versions.
     */
    @Override
    public void publish() {
        var modrinth = extension.getModrinth();
        var token = modrinth.getToken().trim();
        var projectId = modrinth.getProjectId().trim();
        var files = extension.getFiles();

        var dependecyList = modrinth.getDependencies().stream().map(
                dep -> DependencyMetadata.builder()
                        .projectId(dep.getProjectId())
                        .dependencyType(dep.getDependencyType())
                        .build()
        ).toList();

        var validGameVersions = fetchGameVersions().stream()
                .map(GameVersionTag::version).filter(extension.getGameVersions()::contains).toList();

        if (validGameVersions.isEmpty()) {
            project.getLogger().error("No valid game versions found. Please check your game versions in the configuration.");
            return;
        }

        var releaseType = extension.getReleaseType();
        if (!Constants.VALID_RELEASE_TYPE.contains(releaseType)) {
            project.getLogger().error("Invalid release type. Please check your configuration.");
            return;
        }

        if (!VALID_STATUS.contains(modrinth.getStatus())) {
            project.getLogger().error("Invalid status. Please check your configuration.");
            return;
        }

        var filePartNames = IntStream.range(0, files.getFiles().size())
                .mapToObj(i -> i == 0 ? "file" : "file_" + i)
                .toList();

        var metadata = modrinthMetadata(projectId, validGameVersions, releaseType, modrinth, dependecyList, filePartNames);
        project.getLogger().lifecycle("Metadata JSON: " + GSON.toJson(metadata));

        project.getLogger().lifecycle("Publishing to Modrinth...");
        publishingToModrinth(metadata, files, filePartNames, token);
    }

    /**
     * Sends a multipart upload request to Modrinth containing metadata and all mod files.
     *
     * @param metadata       the JSON metadata describing this version
     * @param files          the mod files to upload
     * @param filePartNames  generated field names for each file part
     * @param token          Modrinth API authorization token
     * @throws FailedFileUploadException if the upload request fails
     */
    private void publishingToModrinth(ModrinthMetadata metadata,
                                      ConfigurableFileCollection files,
                                      List<String> filePartNames,
                                      String token) {

        var multipartBuilder = new MultipartBody.Builder().setType(MultipartBody.FORM);

        multipartBuilder.addFormDataPart(
                Constants.DATA,
                null,
                RequestBody.create(GSON.toJson(metadata), MediaType.parse(Constants.MEDIA_TYPE_JSON))
        );

        files.getFiles().forEach(file ->
                multipartBuilder.addFormDataPart(
                        filePartNames.get(new ArrayList<>(files.getFiles()).indexOf(file)),
                        file.getName(),
                        RequestBody.create(file, MediaType.parse(Constants.MEDIA_TYPE_JAR))
                )
        );

        var requestBody = multipartBuilder.build();
        var request = new Request.Builder()
                .url(UPLOAD_URL)
                .header("Authorization", token)
                .header("User-Agent", Constants.USER_AGENT)
                .post(requestBody)
                .build();

        try (var response = client.newCall(request).execute()) {
            var body = response.body().string();

            if (!response.isSuccessful()) {
                project.getLogger().lifecycle("Upload failed. Response body: " + body);

                throw new FailedFileUploadException(
                        "Failed to upload mod to Modrinth: " +
                                response.code() + " - " +
                                response.message() + " - BODY: " + body
                );
            }

            project.getLogger().lifecycle("Successfully uploaded mod to Modrinth!");
        } catch (IOException e) {
            throw new FailedFileUploadException("Failed to upload mod to Modrinth: " + e.getMessage());
        }
    }

    /**
     * Builds a {@link ModrinthMetadata} instance from configuration values.
     *
     * @param projectId        the Modrinth project ID
     * @param validGameVersions game versions confirmed valid by Modrinth
     * @param releaseType      version release channel (e.g., "release", "beta", "alpha")
     * @param modrinth         Modrinth-specific configuration
     * @param dependecyList    dependencies declared for this version
     * @param filePartNames    names for multipart file sections
     * @return fully populated {@code ModrinthMetadata}
     */
    private ModrinthMetadata modrinthMetadata(String projectId,
                                              List<String> validGameVersions,
                                              String releaseType,
                                              ModrinthConfig modrinth,
                                              List<DependencyMetadata> dependecyList,
                                              List<String> filePartNames) {

        return ModrinthMetadata.builder()
                .projectId(projectId)
                .name(extension.getDisplayName())
                .versionNumber(extension.getVersion())
                .changelog(extension.getChangelog())
                .gameVersions(validGameVersions)
                .loaders(extension.getLoaders())
                .releaseChannel(releaseType)
                .featured(modrinth.isFeatured())
                .status(modrinth.getStatus())
                .dependencies(dependecyList)
                .fileParts(filePartNames)
                .build();
    }

    /**
     * Fetches the full list of game version tags from Modrinth.
     *
     * @return list of {@link GameVersionTag} objects
     * @throws FailedFetchGameVersionsException if Modrinth returns an error or the request fails
     */
    private List<GameVersionTag> fetchGameVersions() {
        var request = new Request.Builder()
                .url(GAME_VERSION_URL)
                .get()
                .build();

        try (var response = client.newCall(request).execute()) {
            if (!response.isSuccessful())
                throw new FailedFetchGameVersionsException(
                        "Failed to fetch game versions: " +
                                response.code() + " - " + response.message()
                );

            var responseBody = response.body().string();
            return GSON.fromJson(responseBody, new TypeToken<List<GameVersionTag>>() {}.getType());
        } catch (IOException e) {
            throw new FailedFetchGameVersionsException("Failed to fetch game versions" + e.getMessage());
        }
    }
}

File: src/main/kotlin/io/github/smootheez/modrinth/ModrinthPublisher.kt
// ModrinthPublisher: Publishes mod artifacts to Modrinth via API
package io.github.smootheez.modrinth

import io.github.smootheez.McModPublisherExtension
import io.github.smootheez.PublisherConfig
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import org.gradle.api.Project
import java.io.File

// API endpoints used for publishing and retrieving game versions
private const val UPLOAD_URL = "https://api.modrinth.com/v2/version"
private const val GAME_VERSION_URL = "https://api.modrinth.com/v2/tag/game_version"

// Data class used to deserialize game version data from Modrinth
@Serializable
private data class GameVersionTag(
    val version: String,
    @SerialName("version_type") val versionType: String,
    val date: String,
    val major: Boolean
)

class ModrinthPublisher(
    private val project: Project,
    private val extension: McModPublisherExtension,
    private val client: OkHttpClient = OkHttpClient()
) {
    // JSON parser with unknown field tolerance
    private val json = Json { ignoreUnknownKeys = true }

    fun publish() {
        val modrinth = extension.modrinth
        val token = modrinth.token?.trim()
        val projectId = modrinth.projectId?.trim()
        val files = extension.files.files

        project.logger.lifecycle("🚀  Publishing to Modrinth...")

        // ✅ Validate required fields
        val missing = buildList {
            if (token.isNullOrBlank()) add("Modrinth token")
            if (projectId.isNullOrBlank()) add("Project ID")
            if (files.isEmpty()) add("Files to upload")
        }

        if (missing.isNotEmpty()) {
            project.logger.error("❌ Missing required configuration: ${missing.joinToString(", ")}. Aborting.")
            return
        }

        project.logger.lifecycle("✅ Found ${files.size} file(s) for upload.")
        project.logger.lifecycle("✅ All required configuration is set.")

        // Fetch allowed Minecraft game versions from Modrinth API
        project.logger.lifecycle("⏳ Fetching valid game versions from Modrinth...")
        val validGameVersions = fetchValidGameVersions()
        if (validGameVersions.isEmpty()) {
            project.logger.error("❌ Failed to fetch valid game versions from Modrinth. Aborting.")
            return
        }
        project.logger.lifecycle("✅ Retrieved ${validGameVersions.size} valid game versions.")

        // Validate user configured game versions
        val userGameVersions = extension.gameVersions.map { it.trim() }
        val invalidVersions = userGameVersions.filterNot { it in validGameVersions }

        if (invalidVersions.isNotEmpty()) {
            project.logger.error("❌ Invalid Minecraft version(s): ${invalidVersions.joinToString()}")
            project.logger.error("  ➜ Check available versions on Modrinth or update your configuration.")
            return
        }
        project.logger.lifecycle("✅ Game versions validated: ${userGameVersions.joinToString()}")

        // Ensure release type is valid (release / beta / alpha)
        val releaseType = extension.releaseType.lowercase()
        if (releaseType !in PublisherConfig.VALID_RELEASE_TYPES) {
            project.logger.error("❌ Invalid release type: '$releaseType'")
            project.logger.error("  ➜ Allowed values: release, beta, alpha.")
            return
        }
        project.logger.lifecycle("✅ Release type validated: $releaseType")

        // Build mod metadata payload
        project.logger.lifecycle("⏳ Preparing metadata...")

        val dependencyList = modrinth.dependencies.asMap.values.map {
            DependencyMetadata(
                projectId = it.projectId,
                dependencyType = it.dependencyType
            )
        }

        // Match files to multipart field names
        val filePartNames = files.mapIndexed { index, _ ->
            if (index == 0) "file" else "file_$index"
        }

        // Metadata sent as JSON in the multipart request
        val metadata = ModrinthMetadata(
            projectId = projectId!!,
            name = extension.displayName,
            versionNumber = extension.version ?: project.version.toString(),
            changelog = extension.changelog,
            gameVersions = userGameVersions,
            loaders = extension.loaders,
            releaseChannel = releaseType,
            featured = modrinth.featured,
            status = modrinth.status,
            dependencies = dependencyList,
            fileParts = filePartNames
        )

        val jsonMetadata = json.encodeToString(metadata)
        project.logger.lifecycle("📦 Metadata prepared: $jsonMetadata")

        project.logger.lifecycle("⏳ Uploading to Modrinth...")

        publishingToModrinth(jsonMetadata, files, filePartNames, token!!)
    }

    private fun publishingToModrinth(
        jsonMetadata: String,
        files: Set<File>,
        filePartNames: List<String>,
        token: String
    ) {
        val multipartBuilder = MultipartBody.Builder().setType(MultipartBody.FORM)
        multipartBuilder.addFormDataPart(
            "data",
            null,
            jsonMetadata.toRequestBody(PublisherConfig.MEDIA_TYPE_JSON.toMediaType())
        )

        // Attach all files to the multipart request
        files.forEachIndexed { index, file ->
            val partName = filePartNames[index]
            project.logger.lifecycle("📁 Adding file part: $partName (${file.name})")
            multipartBuilder.addFormDataPart(
                partName,
                file.name,
                file.asRequestBody(PublisherConfig.MEDIA_TYPE_JAR.toMediaType())
            )
        }

        try {
            val requestBody = multipartBuilder.build()
            val request = Request.Builder()
                .url(UPLOAD_URL)
                .addHeader("User-Agent", PublisherConfig.USER_AGENT)
                .addHeader("Authorization", token)
                .post(requestBody)
                .build()

            // Execute the request and handle result
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    val body = response.body.string()
                    project.logger.error("❌ Upload failed: ${response.code} ${response.message}")
                    if (body.isNotBlank()) {
                        project.logger.error("Response body: $body")
                    }
                } else {
                    project.logger.lifecycle("✅ Successfully uploaded to Modrinth!")
                    val responseBody = response.body.string()
                    if (responseBody.isNotBlank()) {
                        project.logger.lifecycle("📨 Response: $responseBody")
                    }
                }
            }
        } catch (e: Exception) {
            project.logger.error("❌ Upload to Modrinth failed due to network or I/O error.")
            project.logger.error("   ➜ ${e::class.simpleName}: ${e.message}")
        }
    }

    private fun fetchValidGameVersions(): Set<String> {
        return try {
            val request = Request.Builder()
                .url(GAME_VERSION_URL)
                .addHeader("User-Agent", PublisherConfig.USER_AGENT)
                .build()

            // Request valid game versions and parse JSON response
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    return emptySet()
                }

                val body = response.body.string()
                if (body.isBlank()) return emptySet()

                val parsed = json.decodeFromString<List<GameVersionTag>>(body)
                parsed.map { it.version }.toSet()
            }
        } catch (e: Exception) {
            project.logger.error("❌ Error fetching game versions: ${e.message}")
            emptySet()
        }
    }
}